ProcessList.c

```diff
@@ -111,7 +114,90 @@ int ProcessList_size(ProcessList* this) {
    return (Vector_size(this->processes));
 }
 
-static void ProcessList_buildTree(ProcessList* this, pid_t pid, int level, int indent, int direction, bool show) {
+static void ProcessList_updateTreeSetLayer(ProcessList* this, Hashtable* newTree, unsigned int left_bound, unsigned int right_bound, unsigned int deep, unsigned int left, unsigned int right, unsigned int* index, unsigned int* tree_index, int indent) {
+   // it's guaranted that layer_size is enough space
+   // but most likely it needs less
+   int layer_size = (right - left) / 2;
+   Vector* layer = Vector_new(this->processes->type, false, layer_size);
+
+   for (unsigned int i = left_bound; i < right_bound; i++) {
+      Process* proc = (Process*)Hashtable_get(this->treeSet, i);
+      if (proc->tree_depth == deep && proc->tree_left > left && proc->tree_right < right) {
+         if (Vector_size(layer) > 0) {
+            // use as tmp variable
+            ((Process*)Vector_get(layer, Vector_size(layer)-1))->tree_depth = proc->tree_index; 
```

Comment:
I am not as provichient in C :disappointed: . 
I could create a variable which I did instead of macros :smile: is there any benefit in macros?

Reply:
Depends on the situation. In this case the macro would have been merely a short-hand for that long expression:

```c
#define LastProcess ((Process*)Vector_get(layer, Vector_size(layer)-1))
// ...
            // use as tmp variable
            LastProcess->tree_depth = proc->tree_index;
// ...
#undef LastProcess
```

But once there is a certain amount of repetition with slight variation involved, this can become [quite handy](https://github.com/htop-dev/htop/pull/229/files#r512234066):
```c
#define resolve(symbolname) do {\
      *(void **)(&sym_##symbolname) = dlsym(dlopenHandle, #symbolname); \
      if (!sym_##symbolname || dlerror() != NULL) \
         goto dlfailure;

      resolve(sd_bus_open_system);
      resolve(sd_bus_get_property_string);
      resolve(sd_bus_get_property_trivial);
      resolve(sd_bus_unref);

#undef resolve
```

Downside with macros usually is that you create repeated code that the optimizer needs to find its way around, whereas variables usually bear no additional runtime cost. One advantage with macros can be, that if written properly (and depending on what you need to do) you can write them so they work as lvalues (and thus can be assigned to) even through several layers of abstraction. For variables you'd have to handle the writing back through the abstraction yourself (depending on what the variable contains. If you can show the pointer for that `vector_get` to be constant you throughout its use here, a variable will be the better solution, if the last element in the vector may change (e.g. by adding/removing items) a macro would be preferred.